lib/block-supports/duotone.php

```diff
@@ -550,6 +541,81 @@ static function () use ( $filter_svg, $selector ) {
 	);
 }
 
+
+
+/**
+ * Outputs all necessary SVG for duotone filters and CSS for classic themes.
+ */
+add_action(
+	'wp_footer',
+	static function () {
+
+		foreach ( WP_Duotone_Gutenberg::$output as $filter_data ) {
+
+			// SVG will be output on the page later.
+			$filter_svg = gutenberg_get_duotone_filter_svg( $filter_data );
+
+			echo $filter_svg;
+
+			// This is for classic themes - in block themes, the CSS is added in the head via wp_add_inline_style in the wp_enqueue_scripts action.
+
+			if ( ! wp_is_block_theme() ) {
+				wp_add_inline_style( 'core-block-supports', 'body{' . WP_Duotone_Gutenberg::get_css_custom_property_declaration( $filter_data ) . '}' );
+			}
+
+			global $is_safari;
+			if ( $is_safari ) {
+				duotone_safari_rerender_hack( $filter_data['selector'] );
+			}
+		}
+	}
+);
+
+/**
+ * Appends the used global style duotone filter CSS Vars to the inline global styles CSS
+ */
+add_action(
+	'wp_enqueue_scripts',
+	static function() {
+
+		if ( empty( WP_Duotone_Gutenberg::$output ) ) {
+			return;
+		}
+
+		$duotone_css_vars = '';
+
+		foreach ( WP_Duotone_Gutenberg::$output as $filter_data ) {
+			if ( ! array_key_exists( $filter_data['slug'], WP_Duotone_Gutenberg::$global_styles_presets ) ) {
+				continue;
+			}
+
+			$duotone_css_vars .= WP_Duotone_Gutenberg::get_css_custom_property_declaration( $filter_data );
+		}
+
+		if ( ! empty( $duotone_css_vars ) ) {
+			wp_add_inline_style( 'global-styles', 'body{' . $duotone_css_vars . '}' );
+		}
+	},
+	11
+);
+
+/**
+ * Safari renders elements incorrectly on first paint when the SVG filter comes after the content that it is filtering,
+ * so we force a repaint with a WebKit hack which solves the issue.
+ *
+ * @param string $selector The selector to apply the hack for.
+ */
+function duotone_safari_rerender_hack( $selector ) {
+	/*
+	 * Simply accessing el.offsetHeight flushes layout and style
+	 * changes in WebKit without having to wait for setTimeout.
+	 */
+	printf(
+		'<script>( function() { var el = document.querySelector( %s ); var display = el.style.display; el.style.display = "none"; el.offsetHeight; el.style.display = display; } )();</script>',
+		wp_json_encode( $selector )
+	);
+}
```

Comment:
This needs to move into `WP_Duotone_Gutenberg`.

Reply:
As a follow-up, I think it'd be good to run this all at once by passing everything from WP_Duotone_Gutenberg::$output and having one <script> tag that handles rerendering all of them.